<a href="https://colab.research.google.com/github/LuisPeMoraRod/AI-Laboratories/blob/main/Lab6_LogisticRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
from math import exp as e

### Ejercicio 1

El conjunto de datos se importa desde un repositorio de GitHub.

El conjunto de datos contiene atributos sobre las cuentas de seguros de un grupo de 1339 ciudadanos estadounidenses. Los atributos del conjunto de datos incluyen: edad, sexo, índice de masa corporal, cantidad de hijos, si es fumador/a, región, cargos.

Este conjunto de datos se va a utilizar para crear un modelo que prediga si una persona es fumadora o no, en función del resto de los atributos.




In [2]:
url = 'https://raw.githubusercontent.com/LuisPeMoraRod/AI-Laboratories/logistic_reg_exercise/logistic_reg_lab_data.csv'
df = pd.read_csv(url) #store in pandas dataframe
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

### Ejercicio 2

In [9]:
#parse data frame to make all atributes numeric
df['sex'] = df['sex'].astype('category')
df['sex'] = df['sex'].cat.codes

df['region'] = df['region'].astype('category')
df['region'] = df['region'].cat.codes

df['smoker'] = df['smoker'].astype('category')
df['smoker'] = df['smoker'].cat.codes

# separate X variables from y
X = df.drop(columns = 'smoker')
y = df['smoker']

#define training and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state = 0)

#use sklean linear regression model to get linear regression equation
linear_regression = LinearRegression()
linear_regression.fit(X_train, y_train)
b = linear_regression.intercept_
coeffs = linear_regression.coef_
print('Linear regression:')
print(f'b = {b}')
print(f'Coefficients = {coeffs}')

X_train = np.array(X_train)

def predict(x_data: np.array, coeffs: list, b: float, threshold: float) -> np.array:
  coeffs = np.array(coeffs)
  rows = len(x_data)
  cols = len(coeffs)
  ones = np.ones((rows, cols))
  coeffs = np.multiply(ones, coeffs)

  # compute logit for every entry
  logits = np.multiply(x_data, coeffs) # multiply every variable with the coefficient computed from linear regression
  logits = np.sum(logits, axis=1) # add every variable
  logits = logits + b # add intercept value

  # apply sigmoid to every logit element
  def sigmoid(num: float):
    sigmoid = 1/(1+e(-num))
    if (sigmoid >= threshold): 
      return 1
    return 0
  
  applySigmoid = np.vectorize(sigmoid)
  y_pred = applySigmoid(logits)
  return y_pred

threshold = 0.6 # threshold used to separate between 0s and 1s
y_pred = predict(X_train, coeffs, b, threshold)
print(f'\nPredictions for training data (smokers):\n{y_pred}')
print(f'\nReal training data (smokers):\n{y_train.to_numpy()}')

accuracy = accuracy_score(y_train, y_pred)
print(f'\nAccuracy score: {accuracy} %')

y_test_pred = predict(X_test, coeffs, b, threshold)
print(f'\nPredictions for test data (smokers):\n{y_test_pred}')
print(f'\nReal test data (smokers):\n{y_test.to_numpy()}')

accuracy = accuracy_score(y_test, y_test_pred)
print(f'\nAccuracy score: {accuracy} %')

Linear regression:
b = 0.410520512058796
Coefficients = [-7.90211044e-03  1.63808541e-02 -9.72718155e-03 -1.06964489e-02
  7.50706739e-03  2.96921242e-05]

Predictions for training data (smokers):
[0 0 0 ... 0 0 0]

Real training data (smokers):
[0 0 0 ... 0 0 0]

Accuracy score: 0.9491525423728814 %

Predictions for test data (smokers):
[0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0
 1 1 0 0 0 1 1 1 1 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0 1 0
 0 0 0 0 1 1 0 1 0 0 1 1 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 0 1 0
 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 1 1 1 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 0 0 1 1
 0 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1
 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
 0 0 1 1 0 0 0 1 0 1 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0
 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 0 1 0 1 0 0 0 0

### Ejercicio 3
**Incluya paso a paso la derivada parcial de la función de activación del algoritmo,es decir, la derivada parcial de la función logística de acuerdo con sus respectivas variables.**

Para esta derivada, vamos a estar derivando la función $p = \frac{1}{1 + e^{-(a_1 \cdot x_1 + a_2 \cdot x_2 + a_n \cdot x_n + b)}}$

De esta manera tenemos que:

$b = 0.40430152275569914 \\
a_1 = -7.84434847e-03 \\
a_2 = 1.44723798e-02 \\
a_3 = -9.80706018e-03 \\
a_4 = -1.29068881e-02 \\
a_5 = 1.16749447e-02 \\
a_6 = 2.99693056e-05 $ \\

Si consideramos nuestro ejemplo inicial, nuestras variables son: \\
$
x_1 = age \\
x_2 = sex \\
x_3 = bmi \\
x_4 = children \\
x_5 = smoker \\
x_6 = region
$

Para evitar trabajar con numeros muy grandes, vamos a aproximar nuestras constantes a 5 decimales, de esta manera tendriamos siguiente expresión: \\
 $p = \frac{1}{1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)}}$

 Y las derivadas parciales de cada una de sus variables quedarian de la siguiente manera:

$
\frac{∂p}{\partial x_1} = -\frac{0.00784 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2} \\
\frac{∂p}{\partial x_2} = \frac{0.01447 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2} \\
\frac{∂p}{\partial x_3} = -\frac{0.00980 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2}\\
\frac{∂p}{\partial x_4} = \frac{0.01290 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2}\\
\frac{∂p}{\partial x_5} = -\frac{0.01674 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2}\\
\frac{∂p}{\partial x_6} = -\frac{0.00002 \cdot e^{0.00784 \cdot x_1-0.01447 \cdot x_2+0.00980 \cdot x_3+0.01290 \cdot x_4-0.01674 \cdot x_5-0.00002 \cdot x_6-0.40430}}{(1 + e^{(0.00784 \cdot x_1 - 0.01447 \cdot x_2 + 0.00980 \cdot x_3 + 0.01290 \cdot x_4 - 0.01674 \cdot x_5 - 0.00002 \cdot x_6 - 0.40430)})^2}\\
$



### Ejercicio 4
A manera de conclusión, se puede afirmar que el modelo generado a partir del algoritmo de *Regresión Logística* para predecir si una persona es fumadora o no, fue exitoso. Esto puesto a que, al ejecutar las predicciones sobre los datos de entrenamiento (75% del total del set de datos), se obtuvo un porcentaje de precisión del 94,9% con respecto a los datos reales. Posteriormente, se utilizó ese mismo modelo para predecir sobre el restante 25% de los datos, que corresponden a los datos de prueba, los cuales no tuvieron ninguna influencia sobre la generación del modelo. En ese caso, se obtuvo un porcentaje de precisión del 96.1% (ver ejercicio 2); un puntaje aún más favorable que el del set de entrenamiento. Nótese que la cantidad de observaciones total del set de datos es de 1338.

Cabe destacar que, para este modelo se utilizó un valor de umbral de 0.6 para dividir los resultados del operador logístico sobre función de la regresión lineal, en 0s y 1s. A partir de la observación, y un proceso de prueba-error, se determinó que el porcentaje de precisión era mucho mayor si se usaba 0.6 en lugar de 0.5, que suele ser el valor tradicional. Ejecutando el modelo con 0.5 como umbral, se obtuvo un porcentaje de precisión del 47.1% para los datos de entrenamiento y 46.0% para los datos de prueba. Dicho rendimiento es bastante deficiente, razón por la cual se optó por ajustar el valor obteniendo mejorías muy claras.

Otro aspecto a resaltar, es que se demostró la idoneidad de este algoritmo para trabajar con problemas de múltiples variables independientes. En este caso, se procesó un set de datos que contenía las variables de edad, sexo, índice de masa corporal, cantidad de hijos, región de residencia (EEUU), y cargos por seguro, para determinar la condición de si la persona es fumadora o no. Así, un total de 6 variables independientes se utilizaron para hacer una clasificación binaria de la variable dependiente.